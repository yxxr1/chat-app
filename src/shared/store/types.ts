import { CONNECTION_METHODS } from '@const/settings';

export type Chat = {
  id: string;
  name: string;
  messages: (Message | undefined)[];
  joinedCount: number | null;
};

export type Message = {
  id: string;
  text: string | null;
  fromId: string;
  fromName: string;
  date: number;
  service?: number;
  index: number;
};

export type UserSettings = {
  connectionMethod: (typeof CONNECTION_METHODS)[keyof typeof CONNECTION_METHODS];
  theme: 'light' | 'dark';
};

export type User = {
  id: string;
  name: string;
  settings: UserSettings;
};

export interface State {
  isLoading: boolean;
  user: User | null;
  allChats: { [chatId: Chat['id']]: Chat };
  joinedChatsIds: Chat['id'][];
  subscribedChatsIds: Chat['id'][];
  currentChatId: Chat['id'] | null;
}
