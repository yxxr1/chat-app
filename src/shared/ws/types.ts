import { Message, Chat } from '@store/types';

type WSMessage<Type, Payload> = {
  type: Type;
  payload: Payload;
};

export type WatchChats = WSMessage<
  'WATCH_CHATS',
  {
    newChats: Chat[];
    deletedChatsIds: Chat['id'][];
    updatedChats: Chat[];
  }
>;

export type ChatSubscribe = WSMessage<
  'SUBSCRIBED_CHAT',
  {
    chatId: Chat['id'];
    messages: Message[];
  }
>;

export type WSMessageIncoming = ChatSubscribe | WatchChats;

export type WSMessageOutgoing =
  | WSMessage<
      'PUBLISH_MESSAGE',
      {
        chatId: Chat['id'];
        message: Message['text'];
      }
    >
  | WSMessage<
      'SUBSCRIBE_CHAT',
      {
        chatId: Chat['id'];
        lastMessageId: Message['id'] | null;
      }
    >;
