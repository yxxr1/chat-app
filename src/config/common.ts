export const COMMON_CONFIG = {
  API_URL: process.env.API_URL || 'http://localhost:8080',
  WS_URL: process.env.WS_URL || 'ws://localhost:8080/ws',
};
