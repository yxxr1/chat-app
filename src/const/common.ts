export const MESSAGE_SERVICE_TYPES = {
  CHAT_JOINED: 1,
  CHAT_LEFT: 2,
} as const;
