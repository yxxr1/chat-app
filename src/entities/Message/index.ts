export { Message, MessageSkeleton } from './Message';
