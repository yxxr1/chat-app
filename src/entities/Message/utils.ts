import moment from 'moment';
import { Message as MessageType } from '@store/types';

export const getUserTitle = (message: MessageType) => `${message.fromName}(${message.fromId.substr(message.fromId.length - 4)})`;

export const formatDate = (date: number) => moment(date).calendar();
