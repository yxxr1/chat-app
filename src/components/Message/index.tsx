import React, { useMemo } from 'react';
import moment from 'moment';
import { Message as MessageType } from '@store/types';
import { getServiceMessage } from '@utils/common';
import { Message as MessageC, UserTitle, MessageDateTime, ServiceMessage } from './styled';

type Props = {
  message: MessageType;
};

export const Message: React.FC<Props> = ({ message }) => {
  const userTitle = useMemo(() => `${message.fromName}(${message.fromId.substr(message.fromId.length - 4)})`, [message]);

  if (message.service) {
    return <ServiceMessage>{getServiceMessage(message)}</ServiceMessage>;
  }

  return (
    <MessageC>
      <UserTitle>{userTitle}</UserTitle>
      <span>{message.text}</span>
      <MessageDateTime>{moment(message.date).calendar()}</MessageDateTime>
    </MessageC>
  );
};
