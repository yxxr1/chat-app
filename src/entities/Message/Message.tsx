import React, { useMemo, memo } from 'react';
import { Message as MessageType } from '@store/types';
import { getServiceMessage } from '@utils/common';
import { getUserTitle, formatDate } from './utils';
import { Message as MessageC, MessageUser, MessageText, MessageDateTime, ServiceMessage, Skeleton } from './styled';

type Props = {
  message: MessageType;
};

const _Message: React.FC<Props> = ({ message }) =>
  message.service ? (
    <ServiceMessage>{getServiceMessage(message)}</ServiceMessage>
  ) : (
    <MessageC data-testid="message__main-container">
      <MessageUser>{getUserTitle(message)}</MessageUser>
      <MessageText data-testid="message__main-container-text">{message.text}</MessageText>
      <MessageDateTime>{formatDate(message.date)}</MessageDateTime>
    </MessageC>
  );

const _MessageSkeleton: React.FC = () => {
  const widthIndex = useMemo(() => Math.floor(Math.random() * 10), []);

  return <Skeleton widthIndex={widthIndex} size="small" />;
};

export const MessageSkeleton = memo(_MessageSkeleton);
export const Message = memo(_Message);
