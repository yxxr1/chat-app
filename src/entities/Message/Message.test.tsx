import React from 'react';
import { cleanup, screen, render } from '@testing-library/react';
import { Message } from '@entities/Message';
import { Message as MessageType } from '@store/types';
import { renderComponent } from '@test/utils';
import { MESSAGE_SERVICE_TYPES } from '@const/common';
import { THEMES } from '@styles/theme';
import { formatDate, getUserTitle } from './utils';

const message: MessageType = {
  id: '8PwpNK37lsBNdbWlUcswm',
  text: 'test message text',
  fromId: 'y_ZAWNqKsDgRLKLDu2Qne',
  fromName: 'test_user',
  date: new Date('2024-07-14').valueOf(),
  service: 0,
  index: 0,
};
const serviceMessage = { ...message, service: MESSAGE_SERVICE_TYPES.CHAT_LEFT };

describe('Message entity', () => {
  afterEach(() => {
    cleanup();
  });

  test('utils', () => {
    expect(getUserTitle(message)).toEqual('test_user(2Qne)');
    expect(formatDate(message.date)).toEqual('07/14/2024');
  });

  test('renders content', () => {
    render(<Message message={message} />);
    expect(screen.getByTestId('message__main-container')).toMatchSnapshot();
  });

  test('renders content service message', () => {
    render(<Message message={serviceMessage} />);
    expect(screen.getByText('test_user left chat')).toBeInTheDocument();
  });

  test('render non service, theme light', () => {
    renderComponent(<Message message={message} />, 'light');
    expect(screen.getByTestId('message__main-container-text')).toHaveStyle({
      color: THEMES.light['text'],
    });
  });

  test('render non service, theme dark', () => {
    renderComponent(<Message message={message} />, 'dark');
    expect(screen.getByTestId('message__main-container-text')).toHaveStyle({
      color: THEMES.dark['text'],
    });
  });
});
