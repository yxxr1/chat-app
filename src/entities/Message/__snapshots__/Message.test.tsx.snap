// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Message entity renders content 1`] = `
<div
  class="sc-aXZVg hRtgRa"
  data-testid="message__main-container"
>
  <div
    class="sc-gEvEer hsltlJ"
  >
    test_user(2Qne)
  </div>
  <span
    class="sc-eqUAAy lsylT"
    data-testid="message__main-container-text"
  >
    test message text
  </span>
  <div
    class="sc-fqkvVR gBYpkF"
  >
    07/14/2024
  </div>
</div>
`;
